src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala

```diff
@@ -0,0 +1,614 @@
+    case 1 => false
+    case 2 => true
+    case n => throw new IllegalArgumentException(s"Found $n read structures with template bases but expected 1 or 2.")
+  }
+
+  /** The concatenation of all read structures read structure. */
+  private val fullReadStructure: ReadStructure = new ReadStructure(segs=readStructures.flatten, resetOffsets=true)
+
+  /** The quality format converter. */
+  private val solexaQualityConverter: SolexaQualityConverter = SolexaQualityConverter.getSingleton
+
+  /** The sub-reads for the sample barcodes across all read structures. */
+  private val sampleBarcodeSegments = fullReadStructure.sampleBarcode
+
+  /** Get the sample barcode bases across all bases. */
+  private def sampleBarcodeBases(allBases: String): String = {
+    ReadStructure.structureRead(bases=allBases, segments=sampleBarcodeSegments).map(_.bases).mkString
+  }
+
+  /** Get the sample barcode across all bases. */
+  private def sampleBarcode(allBases: String): String = {
+    ReadStructure.structureRead(bases=allBases, segments=sampleBarcodeSegments).map(_.bases).filter(_.nonEmpty).mkString("-")
+  }
+
+  /** The sub-reads for the molecular barcodes across all read structures. */
+  private val molecularBarcodeSegments: Seq[MolecularBarcode] = this.fullReadStructure.molecularBarcode
+
+  /** Get the molecular barcode bases across all bases. */
+  private def molecularBarcode(allBases: String): String = {
+    ReadStructure.structureRead(bases=allBases, segments=molecularBarcodeSegments)
+      .map(_.bases)
+      .filter(_.nonEmpty)
+      .mkString("-")
+  }
+
+    /** Gets the sample ordinal given the bases from all cycles and the number of mismatches between the bases and matched
+      * sample barcode.  If no match is found, the ordinal for the oridinal for the 'unmatched" sample is given. */
+  private def matchSampleBarcode(bases: String): (Int, Int) = {
+    val observedBarcode = sampleBarcodeBases(bases).getBytes
+    val numNoCalls      = observedBarcode.count(base => SequenceUtil.isNoCall(base))
+
+    // Get the best and second best sample barcode matches.
+    val (bestSampleIndex, bestMismatches, secondBestMismatches) = if (numNoCalls <= maxNoCalls) {
+      samples.map { sample =>
+        val expectedBarcode = sample.sampleBarcode
+          .getOrElse(unreachable(s"Sample with id '${sample.sampleId}' did not have a sample barcode"))
+          .barcodeBytes
+        val numMismatches = countMismatches(observedBarcode, expectedBarcode)
+        (sample.sampleOrdinal-1, numMismatches)
+      }
+      .toList.sortBy(_._2).take(2) match {
+        case Nil                              => (-1,           Int.MaxValue, Int.MaxValue)
+        case List(bestTuple)                  => (bestTuple._1, bestTuple._2, Int.MaxValue)
+        case List(bestTuple, secondBestTuple) => (bestTuple._1, bestTuple._2, secondBestTuple._2)
+      }
+    }
+    else {
+      (this.samples.size, Int.MaxValue, Int.MaxValue)
```

Comment:
It would be nice to extract `this.samples.size` into a class level val called `UnmatchedSampleOrdinal` and use that everywhere to make it clear what's going on.